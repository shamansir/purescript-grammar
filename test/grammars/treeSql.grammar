main :- repSep(statement, ws).

statement :- [(selectStmt | createTableStmt), ";"].

selectStmt :- [(manyKW | oneKW), ws, tableName, ws, selection].
selection :- ["{", ws, repSep(selectionElement, commaWS), ws, "}"].
selectionElement :- [columnName, (subSelection | "")].
subSelection :- [":", ws, selectStmt].

createTableStmt :- [createKW, ws, tableKW, ws, tableName, ws, "(", ws, repSep(colSpec, commaWS), ws, ")"].

colSpec :- [columnName, ws, type, ws, (referencesClause | "")].

type :- ("INT" | "TEXT").
referencesClause :- [referencesKW, ws, tableName].

columnName :- (ident | placeholder).
tableName :- (ident | placeholder).

createKW :- "CREATE".
tableKW :- "TABLE".
manyKW :- "MANY".
oneKW :- "ONE".
referencesKW :- "REFERENCES".

ident :- [alpha, repSep((alphaNum | "."), "")].
string :- ["\"", repSep(stringChar, ""), "\""].
stringChar :- (^'"' | ['\\', '"']).
alpha :- ([a-z] | [A-Z] | "_").
num :- [0-9].
alphaNum :- (alpha | num).
ws :- repSep((" "|"\n"), "").
commaWS :- [",", ws].
placeholder :- "???".
