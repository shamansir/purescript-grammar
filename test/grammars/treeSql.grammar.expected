main :- repSep(statement,ws).
alpha :- ([a-z]|[A-Z]|"_").
alphaNum :- (alpha|num).
colSpec :- [columnName,ws,type,ws,(referencesClause|"")].
columnName :- (ident|placeholder).
commaWS :- [",",ws].
createKW :- "CREATE".
createTableStmt :- [createKW,ws,tableKW,ws,tableName,ws,"(",ws,repSep(colSpec,commaWS),ws,")"].
ident :- [alpha,repSep((alphaNum|"."),"")].
manyKW :- "MANY".
num :- [0-9].
oneKW :- "ONE".
placeholder :- "???".
referencesClause :- [referencesKW,ws,tableName].
referencesKW :- "REFERENCES".
selectStmt :- [(manyKW|oneKW),ws,tableName,ws,selection].
selection :- ["{",ws,repSep(selectionElement,commaWS),ws,"}"].
selectionElement :- [columnName,(subSelection|"")].
statement :- [(selectStmt|createTableStmt),";"].
string :- ["\"",repSep(stringChar,""),"\""].
stringChar :- (^'"'|['\\','"']).
subSelection :- [":",ws,selectStmt].
tableKW :- "TABLE".
tableName :- (ident|placeholder).
type :- ("INT"|"TEXT").
ws :- repSep((" "|"\n"),"").